environment/src/main/java/com/sequenceiq/environment/network/service/SubnetIdProvider.java

```diff
@@ -1,27 +1,66 @@
 package com.sequenceiq.environment.network.service;
 
+import static com.sequenceiq.cloudbreak.common.mappable.CloudPlatform.AZURE;
+
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Optional;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.stereotype.Component;
 
 import com.sequenceiq.cloudbreak.cloud.model.CloudSubnet;
+import com.sequenceiq.cloudbreak.common.mappable.CloudPlatform;
 import com.sequenceiq.common.api.type.Tunnel;
 import com.sequenceiq.environment.network.dto.NetworkDto;
 
 @Component
 public class SubnetIdProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SubnetIdProvider.class);
 
-    public String provide(NetworkDto network, Tunnel tunnel) {
-        if (network != null && network.getSubnetIds() != null && !network.getSubnetIds().isEmpty()) {
-            return getSubnetIdByPreferedSubnetType(network, tunnel.useCcm());
-        } else {
+    public String provide(NetworkDto network, Tunnel tunnel, CloudPlatform cloudPlatform) {
+        LOGGER.debug("Choosing subnet, network: {},  platform: {}, tunnel: {}", network, cloudPlatform, tunnel);
+        if (network == null || network.getSubnetIds() == null || network.getSubnetIds().isEmpty() || network.getSubnetMetas().isEmpty()) {
+            LOGGER.debug("Check failed, returning null");
             return null;
         }
+        return getSubnetIdByPreferedSubnetType(network, tunnel.useCcm(), cloudPlatform);
+    }
+
+    private String getSubnetIdByPreferedSubnetType(NetworkDto network, boolean preferPrivate, CloudPlatform cloudPlatform) {
+        List<CloudSubnet> subnetMetas = new ArrayList<>(network.getSubnetMetas().values());
+        return AZURE.equals(cloudPlatform)
+                ? subnetMetas.get(0).getId()
+                : getSubnetIdByPreferedSubnetTypeAws(subnetMetas, preferPrivate);
+    }
+
+    private String getSubnetIdByPreferedSubnetTypeAws(List<CloudSubnet> subnetMetas, boolean preferPrivate) {
+        LOGGER.debug("Choosing aws subnet, preferring private: {}, available", preferPrivate);
+        Optional<CloudSubnet> foundCloudSubnet = preferPrivate ? tryGetOnePrivateSubnet(subnetMetas) : tryGetOnePublicSubnet(subnetMetas);
+        if (foundCloudSubnet.isEmpty()) {
+            LOGGER.debug("Choosing subnet alternate strategy");
+            // This is just trying the reverse option, i.e. if we preferred private but there was no private then let's look for a suitable public subnet
+            foundCloudSubnet = preferPrivate ? tryGetOnePublicSubnet(subnetMetas) : tryGetOnePrivateSubnet(subnetMetas);
+        }
+        if (foundCloudSubnet.isEmpty()) {
+            LOGGER.debug("Choosing subnet, fallback to any subnet");
+            foundCloudSubnet = Optional.of(subnetMetas.get(0));
+        }
```

Comment:
it would reflect the logic more if you move this if into the upper one

Reply:
I see. Actually I like to avoid many indentations, they are hard to read. Also, this is yet another step to try to get a subnet, different from the previous one.

Reply:
also can be moved to a separate method called like `fallbackToAny`